h/services/bulk_executor/_actions.py

```diff
@@ -79,6 +84,46 @@ def _execute_statement(self, stmt):
         return result
 
 
+class GroupUpsertAction(DBAction):
+    """Perform a bulk group upsert."""
+
+    def execute(self, batch, effective_user_id=None, **_):
+        if effective_user_id is None:
+            raise CommandSequenceError(
+                "Effective user must be configured before upserting groups"
+            )
+
+        # Check that we can actually process this batch
+        self._check_upsert_queries(
+            batch, expected_keys=["authority", "authority_provided_id"]
+        )
+
+        # Prep the query
+        values = [command.body.attributes for command in batch]
+        for value in values:
+            value["creator_id"] = effective_user_id
+
+        stmt = insert(Group).values(values)
+        stmt = stmt.on_conflict_do_update(
+            index_elements=["authority", "authority_provided_id"],
+            set_={"name": stmt.excluded.name},
+        ).returning(Group.id, Group.authority, Group.authority_provided_id)
```

Comment:
You can't one-liner this as the set clause is specified using values on the `stmt`.